File: LeetCode/Explore/TopInterview/Blind75/Array/153_find_minimum in_rotated_sorted_array.cpp
// one pass binary search
// Complexity Analysis
// Time Complexity : Same as Binary Search O(\log N)O(logN)
// Space Complexity : O(1)O(1)

class Solution {
public:
  int findMin(vector<int>& nums) {
    int pivot, l = 0, r = nums.size() - 1;
    while (l < r) {
      pivot = l + (r - l) / 2;
      if (nums[pivot] > nums[r]) {
        l = pivot + 1;
      }
      else {
        r = pivot;
      }
    }
    return nums[l];
  }
};

File: LeetCode/Explore/TopInterview/Blind75/Array/153_find_minimum_in_rotated_sorted_array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
      l, r = 0, len(nums) - 1
      
      while l < r:
        pivot = l + (r - l) // 2
        if nums[pivot] < nums[r]:
          r = pivot
        else:
          l = pivot + 1
          
      return nums[l]
